### PR TITLE
fix: [DHIS2-17054] handle empty expression value

### PR DESCRIPTION
**Pull request by simonadomnisoru**
[DHIS2-17054](https://dhis2.atlassian.net/browse/DHIS2-17054)

**Tech summary**
- handle empty expression value in `processDisplayKeyValuePair`
- added a unit test for the `undefined` value use case
- small CSS adjustments

[DHIS2-17054]: https://dhis2.atlassian.net/browse/DHIS2-17054?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js [hunk 1]
```diff
@@ -239,13 +239,17 @@ export function getRulesEffectsProcessor(
     }
 
     function processDisplayKeyValuePair(effect: ProgramRuleEffect): any {
+        const { data } = effect;
+        const dataToString = value => (typeof value == 'number' ? numberToString(value) : String(value));
+        const value = data !== undefined ? dataToString(data) : '';
```
Comment (superskip) [inline packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js:244]:
```suggestion
        const data = effect.data || '';
```

File: packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js [hunk 2]
```diff
@@ -239,13 +239,17 @@ export function getRulesEffectsProcessor(
...
+
         return {
             type: effectActions.DISPLAY_KEY_VALUE_PAIR,
             id: effect.location,
             displayKeyValuePair: {
                 id: effect.id,
                 key: effect.displayContent,
-                value: typeof effect.data == 'number' ? numberToString(effect.data) : String(effect.data),
+                value,
```
Comment (superskip) [inline packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js:252]:
```suggestion
                value: typeof data === 'number' ? numberToString(data) : String(data),
```

**Review by superskip — COMMENTED** [review]
I found a small simplification, posted as comments. Regarding the css changes, will the bottom margin be constant 0 now? How did it work previously (what does '& + &' mean)?

File: packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js (hunk 1 above)
Comment (superskip) [reply packages/rules-engine/src/processors/rulesEffectsProcessor/rulesEffectsProcessor.js:244]: Hold on, maybe `effect.data` could be `false`.. In that case this won't work.

Revised suggestion:
```suggestion
        const data = effect.data !== undefined ? effect.data : '';
```

**Comment by simonadomnisoru** [comment]
> I found a small simplification, posted as comments. Regarding the css changes, will the bottom margin be constant 0 now? How did it work previously (what does '& + &' mean)?

Hey,
I applied the small simplifications you suggested. As for the bottom margins, yes, they are all constant now as requested in the ticket. 
From what I understand from [next sibling combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Next-sibling_combinator) and [nesting selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector#appending_the_nesting_selector), `& + &` compiles to `.listItem + .listItem` and matches from the second `.listItem` class until the last one. This means that before the first listItem had a different bottom margin than the second and the rest ones.
Thanks for the feedback!

**Comment by superskip** [comment]
> From what I understand from [next sibling combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Next-sibling_combinator) and [nesting selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector#appending_the_nesting_selector), & + & compiles to .listItem + .listItem and matches from the second .listItem class until the last one. This means that before the first listItem had a different bottom margin than the second and the rest ones.

Thanks for the explanation :)

@CooperJoe Do you recall what you tried to achieve by using the `& + &` selector?

**Review by superskip — APPROVED** [review]

**Review by geethaalwan — CHANGES_REQUESTED** [review]
Tested successfully on 2.42,2.41.1,2.40.4,2.39.6 versions

**Review by geethaalwan — APPROVED** [review]

**Pull request merged**